pkg/tfbridge/walk.go

```diff
@@ -168,3 +284,394 @@ func LookupSchemaInfoPath(schemaPath SchemaPath, schemaInfo *SchemaInfo) *Schema
 		return nil
 	}
 }
+
+// Drill down a path from a SchemaInfo object and find a matching SchemaInfo. If no
+// SchemaInfo exists, it will be created.
+func getOrCreateSchemaInfoPath(schemaPath SchemaPath, schemaInfo *SchemaInfo) *SchemaInfo {
+	if len(schemaPath) == 0 {
+		return schemaInfo
+	}
+
+	switch first := schemaPath[0].(type) {
+	case walk.ElementStep:
+		return getOrCreateSchemaInfoPath(schemaPath[1:], ensureField(&schemaInfo.Elem))
+	case walk.GetAttrStep:
+		return getOrCreateSchemaInfoPath(schemaPath[1:], ensureMapKey(&schemaInfo.Fields, first.Name))
+	default:
+		contract.Failf("Unable to walk SchemaPath segment of type %T", schemaPath[0])
+		return nil
+	}
+}
+
+type VisitInfo struct {
+	Root VisitRoot
+
+	schemaPath SchemaPath
+
+	shimSchema    shim.Schema
+	getSchemaInfo func() *SchemaInfo
+
+	getShimSchemaMap func() shim.SchemaMap
+	getSchemaInfoMap func() map[string]*SchemaInfo
+}
+
+type VisitResult struct {
+	// If the visitor has done something.
+	//
+	// Visits will only be replayed during runtime startup if `HasEffect: true`.
+	HasEffect bool
+}
+
+func (v *VisitInfo) SchemaPath() SchemaPath                         { return v.schemaPath }
+func (v *VisitInfo) ShimSchema() shim.Schema                        { return v.shimSchema }
+func (v *VisitInfo) SchemaInfo() *SchemaInfo                        { return v.getSchemaInfo() }
+func (v *VisitInfo) EnclosingSchemaInfoMap() map[string]*SchemaInfo { return v.getSchemaInfoMap() }
+func (v *VisitInfo) EnclosingSchemaShimMap() shim.SchemaMap         { return v.getShimSchemaMap() }
+
+type VisitRoot interface {
+	PulumiToken() tokens.Type
+
+	isVisitRoot()
+}
+
+// This ensures that we can peform casts from VisitRoot to T for each valid root. Changing
+// this from T to *T is a breaking change.
+var (
+	_ VisitRoot = VisitResourceRoot{}
+	_ VisitRoot = VisitDataSourceRoot{}
+	_ VisitRoot = VisitConfigRoot{}
+)
+
+type VisitResourceRoot struct {
+	token tokens.Type
+
+	Info *ResourceInfo
+
+	TfToken string
+}
+
+func (r VisitResourceRoot) PulumiToken() tokens.Type { return r.token }
+func (r VisitResourceRoot) isVisitRoot()             {}
```

Comment:
`isVisitRoot()` is here only to satisfy the interface, and yes this is a common go idiom. Because `isVisitRoot` is a private function, this means that only types in the `tfbridge` package can implement the `VisitRoot` interface, making `VisitRoot` a "closed" interface. 

I think the naming is idiomatic.